messaging/Controllers/BundlesController.cs

```diff
@@ -31,15 +31,31 @@ public BundlesController(ApplicationDbContext context, IServiceProvider services
         public async Task<ActionResult<Bundle>> GetOutgoingMessageItems(string jurisdictionId, DateTime _since = default(DateTime))
         {
             // TODO only allow the since param in development
-            var messageTasks = _context.OutgoingMessageItems.Where(message => message.CreatedDate >= _since && message.JurisdictionId == jurisdictionId).ToList()
-                                                            .Select(message => System.Threading.Tasks.Task.Run(() => BaseMessage.Parse(message.Message, true)));
+            // if _since is the default value, then apply the retrieved at logic
+            System.Collections.Generic.IEnumerable<System.Threading.Tasks.Task<VRDR.BaseMessage>> messageTasks;
+            if (_since == default(DateTime))
+            {
+                // get all messages that have not yet been retrieved 
+                messageTasks = _context.OutgoingMessageItems.Where(message => message.RetrievedAt == null && message.JurisdictionId == jurisdictionId).ToList()
+                                                            .Select(message => System.Threading.Tasks.Task.Run(() => BaseMessage.Parse(message.Message, true))); 
+            }
+            else
+            {
+                messageTasks = _context.OutgoingMessageItems.Where(message => message.CreatedDate >= _since && message.JurisdictionId == jurisdictionId).ToList()
+                                                            .Select(message => System.Threading.Tasks.Task.Run(() => BaseMessage.Parse(message.Message, true)));            
+            }
             Bundle responseBundle = new Bundle();
             responseBundle.Type = Bundle.BundleType.Searchset;
             responseBundle.Timestamp = DateTime.Now;
             var messages = await System.Threading.Tasks.Task.WhenAll(messageTasks);
+            DateTime retrievedTime = DateTime.UtcNow;
             foreach (var message in messages)
             {
                 responseBundle.AddResourceEntry((Bundle)message, "urn:uuid:" + message.MessageId);
+                // update each outgoing message's RetrievedAt field
+                OutgoingMessageItem msgItem = _context.OutgoingMessageItems.Where(msg => msg.MessageId == message.MessageId).First();

```

Comment:
This will iterate the list again (potentially many times, if there were a lot of updates). Could be a performance hit. Almost certainly premature optimization to try do it differently at this point, but maybe worth adding a comment addressing it in case performance becomes an issue in testing?

As far as an alternate approach; if we held onto a reference to the message after the initial query, and made the relevant changes to that, would it get properly detected/saved by EF?

Reply:
I changed this so now there are two iterations, one over messages and one over the outgoingMessageItems. But the outgoingMessageItems retains a reference from the select and calls a ForEach update on list of items. This prevents querying the database for each message again.